drivers/usb/usbuhci_new/roothub.c

```diff
@@ -0,0 +1,470 @@
+#include "usbuhci.h"
+
+//#define NDEBUG
+#include <debug.h>
+
+VOID
+NTAPI
+UhciRHGetRootHubData(IN PVOID uhciExtension,
+                     IN PVOID rootHubData)
+{
+    PUHCI_EXTENSION UhciExtension = uhciExtension;
+    PUSBPORT_ROOT_HUB_DATA RootHubData = rootHubData;
+    USBPORT_HUB_11_CHARACTERISTICS HubCharacteristics;
+
+    DPRINT("UhciRHGetRootHubData: ...\n");
+
+    HubCharacteristics.AsUSHORT = 0;
+    HubCharacteristics.PowerControlMode = TRUE;
+    HubCharacteristics.NoPowerSwitching = TRUE;
+    HubCharacteristics.OverCurrentProtectionMode = TRUE;
+
+    if (UhciExtension->HcFlavor != UHCI_Piix4)
+    {
+        HubCharacteristics.NoOverCurrentProtection = TRUE;
+    }
+
+    RootHubData->NumberOfPorts = UHCI_NUM_ROOT_HUB_PORTS;
+    RootHubData->HubCharacteristics.Usb11HubCharacteristics = HubCharacteristics;
+    RootHubData->PowerOnToPowerGood = 1;
+    RootHubData->HubControlCurrent = 0;
+}
+
+MPSTATUS
+NTAPI
+UhciRHGetStatus(IN PVOID uhciExtension,
+                IN PUSHORT Status)
+{
+    DPRINT("UhciRHGetStatus: ...\n");
+    *Status = 1;
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+UhciRHGetPortStatus(IN PVOID uhciExtension,
+                    IN USHORT Port,
+                    IN PUSB_PORT_STATUS_AND_CHANGE PortStatus)
+{
+    PUHCI_EXTENSION UhciExtension = uhciExtension;
+    PUHCI_HW_REGISTERS BaseRegister;
+    PUSHORT PortControlRegister;
+    UHCI_PORT_STATUS_CONTROL PortControl;
+    ULONG port;
+    USB_20_PORT_STATUS portStatus;
+    USB_20_PORT_CHANGE portChange;
+
+    //DPRINT("UhciRHGetPortStatus: ...\n");
+
+    BaseRegister = UhciExtension->BaseRegister;
+    PortControlRegister = &BaseRegister->PortControl[Port-1].AsUSHORT;
+    PortControl.AsUSHORT = READ_PORT_USHORT(PortControlRegister);
+
+    portStatus.AsUshort16 = 0;
+    portChange.AsUshort16 = 0;
+
+    portStatus.CurrentConnectStatus = PortControl.CurrentConnectStatus;
+    portStatus.PortEnabledDisabled = PortControl.PortEnabledDisabled;
+
+    if (PortControl.Suspend == TRUE &&
+        PortControl.PortEnabledDisabled == TRUE)
+    {
+        portStatus.Suspend = TRUE;
+    }
+    else
+    {
+        portStatus.Suspend = FALSE;
+    }
+
+    // FIXME HcFlavor in usbport
+    //if (UhciExtension->HcFlavor == UHCI_Piix4)
+    if (TRUE)
+    {
+        portStatus.OverCurrent = PortControl.Reserved2 & 1;
+        portStatus.PortPower = (~PortControl.Reserved2 & 1);
+        portChange.OverCurrentIndicatorChange = (PortControl.Reserved2 & 2) != 0;
+    }
+    else
+    {
+        portStatus.OverCurrent = FALSE;
+        portStatus.PortPower = TRUE;
+        portChange.OverCurrentIndicatorChange = FALSE;
+    }
+
+    portStatus.HighSpeedDeviceAttached = FALSE;
+
+    portStatus.Reset = PortControl.PortReset;
+    portStatus.LowSpeedDeviceAttached = PortControl.LowSpeedDevice;
+    portChange.ConnectStatusChange = PortControl.ConnectStatusChange;
+
+    port = 1 << (Port - 1);
```

Comment:
`portBit` or something would be more descriptive.

Reply:
Ranamed